File: chap14_respiration/compartmental_O.m
%  -------------------------------------------------------------------
%
%  The compartmental model for oxygen uptake by the lungs.
%
%   For Chapter 14, Section 14.4.5 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
%
%   Written by James Keener and James Sneyd.
%
%  -------------------------------------------------------------------

function compartmental_O

clear all
close all
clc

set(0,                           ...
   'defaultaxesfontsize', 20,   ...
   'defaultaxeslinewidth', 1.0, ...
   'defaultlinelinewidth', 2.0,...
   'defaultlinemarkersize',12);

global p

p.sigma = 1.4;
p.KO = 30*p.sigma;
p.THb = 2200;
p.RT = 1.7e-2;   % units of uM
p.D =  32.5;
p.PO =  150;  % air partial pressure
p.gamma = 37;

% solve the ODE for TO
init = 6000;
tspan = linspace(0,10,100);
[T,Y] = ode15s(@Tode,tspan,init);
for j = 1:length(T)
    [O(j),Otilde(j)] = getO_Otilde(Y(j));
end

figure(1)
    plot(T,Y)
    xlabel('time (nondimensional)')
    ylabel('[O_2] in Body')

% the target is Pa = 104, Pv = 40
Oa = 104*p.sigma;
conc(Oa)

figure(2)
    plot(T, O/p.sigma,'r', T , Otilde/p.sigma,'b')
    title('O_2 Partial Pressures')
    xlabel('time (dimensionless')
    O(end)/p.sigma
    Otilde(end)/p.sigma

end % of main



%% ODE for TO
function out = Tode(t,x)
    TO = x;
    global p
    [O,Otilde] = getO_Otilde(TO);
    out = p.D*(p.sigma*p.PO - Otilde) - p.gamma*O;
end

%% get O and Otilde as functions of TO
function [O,Otilde] = getO_Otilde(TO)
    global p
    options = optimset('Display','off');
    O = fsolve(@(x) conc(x)-TO,20,options);
    Otilde = fsolve(@(x) p.D*(p.sigma*p.PO - x) + TO - conc(x),20,options);
end

%% oxygen saturation function
function out=conc(O)
    global p
    out = O + 4*p.THb*O.^4./(p.KO^4 + O.^4);
end

File: chap14_respiration/compartmental_O.py

#   -------------------------------------------------------------------
# 
#   The compartmental model for oxygen uptake by the lungs.
# 
#    For Chapter 14, Section 14.4.5 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
# 
#    Written by James Keener and James Sneyd.
# 
#   -------------------------------------------------------------------

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import odeint
from scipy.optimize import fsolve

def Tode(TO, t):
    O, Otilde = get_O_Otilde(TO)
    return D * (sigma * PO - Otilde) - gamma * O


def get_O_Otilde(TO):
    O = fsolve(lambda x: conc(x) - TO, 20)
    Otilde = fsolve(lambda x: D * (sigma * PO - x) + TO - conc(x), 20)
    return O[0], Otilde[0]

def conc(O):
    return O + 4 * THb * O**4 / (KO**4 + O**4)


# Define global parameters
sigma = 1.4
KO = 30 * sigma
THb = 2200
RT = 1.7e-2  # units of uM
D = 32.5
PO = 150  # air partial pressure
gamma = 37

# Solve the ODE for TO
init = 6000
tspan = np.linspace(0, 10, 100)


sol = odeint(Tode, init, tspan)
O = []
Otilde = []
for TO in sol:
    O_val, Otilde_val = get_O_Otilde(TO)
    O.append(O_val)
    Otilde.append(Otilde_val)

# Plot the results
plt.figure(1)
plt.plot(tspan, sol)
plt.xlabel('Time (nondimensional)')
plt.ylabel('[O_2] in Body')

# Calculate and plot oxygen partial pressures
Oa = 104 * sigma
O_values = np.array(O) / sigma
Otilde_values = np.array(Otilde) / sigma

plt.figure(2)
plt.plot(tspan, O_values, 'r', label='O')
plt.plot(tspan, Otilde_values, 'b', label='Otilde')
plt.title('O_2 Partial Pressures')
plt.xlabel('Time (dimensionless)')
plt.legend()
plt.show()

print(f'O_final: {O[-1] / sigma}')
print(f'Otilde_final: {Otilde[-1] / sigma}')
